lightning/src/routing/router.rs

```diff
@@ -349,63 +394,170 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 			}
 		}
 	}
 
-	while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
-		if pubkey == *our_node_id {
-			let mut res = vec!(dist.remove(&our_node_id).unwrap().3);
-			loop {
-				if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
-					res.last_mut().unwrap().node_features = features.to_context();
-				} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
-					if let Some(node_info) = node.announcement_info.as_ref() {
-						res.last_mut().unwrap().node_features = node_info.features.clone();
+	let mut payment_paths = Vec::new();
+	let mut tries_left = 10;
+	let mut found_amount_msat = 0;
+	// For every MPP start with the same pre-filled data.
+	let prefilled_targets = targets.clone();
+	let prefilled_dist = dist.clone();
+
+	'route_finding: loop {
+		targets = prefilled_targets.clone();
+		dist = prefilled_dist.clone();
+		'path_finding: while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
+			if pubkey == *our_node_id {
+				// TODO: currently pretending that always have blockchain/amount data for channel capacities.
+				// Should handle None better.
+				let mut new_entry = dist.remove(&our_node_id).unwrap();
+				let mut res = vec!(new_entry.3.clone());
+				let mut path_bottleneck_msat = None;
+				let mut cur_path_fees = Vec::new();
+				let mut smallest_channel_id = None;
+				loop {
+					if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
+						res.last_mut().unwrap().node_features = features.to_context();
+					} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
+						if let Some(node_info) = node.announcement_info.as_ref() {
+							res.last_mut().unwrap().node_features = node_info.features.clone();
+						} else {
+							res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						}
 					} else {
-						res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						// We should be able to fill in features for everything except the last
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(res.last().unwrap().pubkey == *target);
+					}
+
+
+					if let Some(current_hop_bottleneck_msat) = new_entry.4 {
+						if let Some(full_path_bottleneck_msat) = path_bottleneck_msat {
+							if current_hop_bottleneck_msat < full_path_bottleneck_msat {
+								path_bottleneck_msat = Some(current_hop_bottleneck_msat);
+								// Store smallest channel per path to avoid hitting it while looking
+								// for another path (per MPP).
+								smallest_channel_id = Some(res.last().unwrap().short_channel_id);
+							}
+						}
+					}
+
+					if let Some(bottleneck_amount_msat) = path_bottleneck_msat {
+						found_amount_msat += bottleneck_amount_msat;
```

Comment:
I'm confused why we're incrementing this in each iteration of the per-hop-in-path loop?

Reply:
this was a bug, fixing it.